From d0bee1cfc3ea9f61b197dbffc8ed429d6e9ca5b4 Mon Sep 17 00:00:00 2001
From: Alexander Shishkin <alexander.shishkin@linux.intel.com>
Date: Tue, 15 Feb 2022 15:13:14 +0200
Subject: [PATCH 1392/1418] efi/gop: Disable for TDX guest

The GOP extensively depends on the input from the host. Given that it is
not useful in TDX guest, disable it.

Signed-off-by: Alexander Shishkin <alexander.shishkin@linux.intel.com>
---
 drivers/firmware/efi/libstub/x86-stub.c | 26 +++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/firmware/efi/libstub/x86-stub.c b/drivers/firmware/efi/libstub/x86-stub.c
index 66bab6e49d0d..9ec058cc78e3 100644
--- a/drivers/firmware/efi/libstub/x86-stub.c
+++ b/drivers/firmware/efi/libstub/x86-stub.c
@@ -16,6 +16,8 @@
 #include <asm/setup.h>
 #include <asm/desc.h>
 #include <asm/boot.h>
+#include <asm/processor.h>
+#include <asm/tdx.h>
 #include <asm/unaccepted_memory.h>
 
 #include "efistub.h"
@@ -27,6 +29,25 @@ const efi_system_table_t *efi_system_table;
 extern u32 image_offset;
 static efi_loaded_image_t *image = NULL;
 
+#ifdef CONFIG_INTEL_TDX_GUEST
+static bool intel_tdx_guest;
+
+static void detect_intel_tdx(void)
+{
+	u32 eax, sig[3];
+
+	cpuid_count(TDX_CPUID_LEAF_ID, 0, &eax, &sig[0], &sig[2], &sig[1]);
+
+	if (memcmp("IntelTDX    ", sig, 12))
+		return;
+
+	intel_tdx_guest = true;
+}
+#else
+#define intel_tdx_guest false
+static void detect_intel_tdx(void) {}
+#endif
+
 static efi_status_t
 preserve_pci_rom_image(efi_pci_io_protocol_t *pci, struct pci_setup_rom **__rom)
 {
@@ -317,6 +338,9 @@ static void setup_graphics(struct boot_params *boot_params)
 	void **gop_handle = NULL;
 	void **uga_handle = NULL;
 
+	if (intel_tdx_guest)
+		return;
+
 	si = &boot_params->screen_info;
 	memset(si, 0, sizeof(*si));
 
@@ -794,6 +818,8 @@ unsigned long efi_main(efi_handle_t handle,
 		image_offset = 0;
 	}
 
+	detect_intel_tdx();
+
 #ifdef CONFIG_CMDLINE_BOOL
 	status = efi_parse_options(CONFIG_CMDLINE);
 	if (status != EFI_SUCCESS) {
-- 
2.31.1

